File: cpp_merge/classified/packages/plugin-sam/src/__tests__/test-utils.cpp
#include "test-utils.h"

std::shared_ptr<IAgentRuntime> createMockRuntime(Partial<std::shared_ptr<IAgentRuntime>> overrides)
{
    return as<std::shared_ptr<IAgentRuntime>>(utils::assign(object{
        object::pair{std:("agentId"), as<std::shared_ptr<UUID>>(std:("00000000-0000-0000-0000-000000000001"))}, 
        object::pair{std:("getSetting"), mock([=](auto key) mutable
        {
            auto settings = object{
                object::pair{std:("SAM_SPEED"), 72}, 
                object::pair{std:("SAM_PITCH"), 64}, 
                object::pair{std:("SAM_THROAT"), 128}, 
                object::pair{std:("SAM_MOUTH"), 128}
            };
            return OR((const_(settings)[key]), (nullptr));
        }
        )}, 
        object::pair{std:("getService"), mock([=]() mutable
        {
            return nullptr;
        }
        )}, 
        object::pair{std:("hasService"), mock([=]() mutable
        {
            return false;
        }
        )}, 
        object::pair{std:("registerService"), mock()}, 
        object::pair{std:("initialize"), mock([=]() mutable
        {
            return Promise->resolve();
        }
        )}, 
        object::pair{std:("stop"), mock([=]() mutable
        {
            return Promise->resolve();
        }
        )}, 
        object::pair{std:("evaluate"), mock([=]() mutable
        {
            return Promise->resolve(nullptr);
        }
        )}, 
        object::pair{std:("processActions"), mock([=]() mutable
        {
            return Promise->resolve();
        }
        )}, 
        object::pair{std:("useModel"), as<any>(mock([=]() mutable
        {
            return Promise->resolve(std:("test response"));
        }
        ))}, 
        object::pair{std:("ensureConnection"), mock([=]() mutable
        {
            return Promise->resolve();
        }
        )}, 
        object::pair{std:("composeState"), mock([=]() mutable
        {
            return Promise->resolve(as<std::shared_ptr<State>>(object{
                object::pair{std:("data"), object{}}, 
                object::pair{std:("values"), object{}}, 
                object::pair{std:("text"), string_empty}
            }));
        }
        )}, 
        object::pair{std:("createMemory"), mock([=]() mutable
        {
            return Promise->resolve(as<std::shared_ptr<UUID>>(std:("test-memory-id")));
        }
        )}, 
        object::pair{std:("actions"), array<any>()}, 
        object::pair{std:("providers"), array<any>()}, 
        object::pair{std:("evaluators"), array<any>()}, 
        object::pair{std:("services"), std::make_shared<Map>()}, 
        object::pair{std:("db"), nullptr}, 
        object::pair{std:("plugins"), array<any>()}, 
        object::pair{std:("routes"), array<any>()}, 
        object::pair{std:("logger"), std:("logger")}, 
        object::pair{std:("character"), object{
            object::pair{std:("name"), std:("Test SAM Agent")}, 
            object::pair{std:("id"), as<std::shared_ptr<UUID>>(std:("00000000-0000-0000-0000-000000000001"))}, 
            object::pair{std:("username"), std:("test-sam-agent")}, 
            object::pair{std:("bio"), array<string>{ std:("Retro voice synthesis agent") }}, 
            object::pair{std:("settings"), object{}}, 
            object::pair{std:("system"), std:("You are a helpful assistant with retro voice synthesis.")}, 
            object::pair{std:("plugins"), array<string>{ std:("@elizaos/plugin-sam") }}
        }}
    }, overrides));
};


void setupLoggerSpies()
{
    spyOn(logger, std:("info"));
    spyOn(logger, std:("error"));
    spyOn(logger, std:("warn"));
    spyOn(logger, std:("debug"));
};


std::shared_ptr<Memory> createMockMemory(Partial<std::shared_ptr<Memory>> overrides)
{
    return utils::assign(object{
        object::pair{std:("id"), as<std::shared_ptr<UUID>>(std:("00000000-0000-0000-0000-000000000002"))}, 
        object::pair{std:("entityId"), as<std::shared_ptr<UUID>>(std:("00000000-0000-0000-0000-000000000003"))}, 
        object::pair{std:("agentId"), as<std::shared_ptr<UUID>>(std:("00000000-0000-0000-0000-000000000001"))}, 
        object::pair{std:("roomId"), as<std::shared_ptr<UUID>>(std:("00000000-0000-0000-0000-000000000004"))}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("test message")}
        }}, 
        object::pair{std:("createdAt"), Date->now()}
    }, overrides);
};


std::shared_ptr<State> createMockState(Partial<std::shared_ptr<State>> overrides)
{
    return utils::assign(object{
        object::pair{std:("data"), object{}}, 
        object::pair{std:("values"), object{}}, 
        object::pair{std:("text"), string_empty}
    }, overrides);
};


std::shared_ptr<Uint8Array> createMockAudioBuffer(double length)
{
    auto buffer = std::make_shared<Uint8Array>(length);
    for (auto i = 0; i < length; i++)
    {
        (*buffer)[i] = Math->floor(Math->random() * 256);
    }
    return buffer;
};


std::shared_ptr<Uint8Array> createMockWAVBuffer(std::shared_ptr<Uint8Array> audioData, double sampleRate)
{
    auto wavHeaderSize = 44;
    auto wavBuffer = std::make_shared<Uint8Array>(wavHeaderSize + audioData->length);
    auto view = std::make_shared<DataView>(wavBuffer->buffer, 0, wavHeaderSize);
    wavBuffer->set(((std::make_shared<TextEncoder>()))->encode(std:("RIFF")), 0);
    view->setUint32(4, 36 + audioData->length, true);
    wavBuffer->set(((std::make_shared<TextEncoder>()))->encode(std:("WAVE")), 8);
    wavBuffer->set(((std::make_shared<TextEncoder>()))->encode(std:("fmt ")), 12);
    view->setUint32(16, 16, true);
    view->setUint16(20, 1, true);
    view->setUint16(22, 1, true);
    view->setUint32(24, sampleRate, true);
    view->setUint32(28, sampleRate, true);
    view->setUint16(32, 1, true);
    view->setUint16(34, 8, true);
    wavBuffer->set(((std::make_shared<TextEncoder>()))->encode(std:("data")), 36);
    view->setUint32(40, audioData->length, true);
    wavBuffer->set(audioData, wavHeaderSize);
    return wavBuffer;
};

File: classified/packages/plugin-sam/src/__tests__/test-utils.ts
import { spyOn, mock } from 'bun:test';
import { logger, type IAgentRuntime, type Memory, type State, type UUID } from '@elizaos/core';

/**
 * Create a mock runtime for testing
 */
export function createMockRuntime(overrides: Partial<IAgentRuntime> = {}): IAgentRuntime {
  return {
    agentId: '00000000-0000-0000-0000-000000000001' as UUID,
    getSetting: mock((key: string) => {
      const settings: Record<string, unknown> = {
        SAM_SPEED: 72,
        SAM_PITCH: 64,
        SAM_THROAT: 128,
        SAM_MOUTH: 128,
      };
      return settings[key] || null;
    }),
    getService: mock(() => null),
    hasService: mock(() => false),
    registerService: mock(),
    initialize: mock(() => Promise.resolve()),
    stop: mock(() => Promise.resolve()),
    evaluate: mock(() => Promise.resolve(null)),
    processActions: mock(() => Promise.resolve()),
    useModel: mock(() => Promise.resolve('test response')) as any,
    ensureConnection: mock(() => Promise.resolve()),
    composeState: mock(() => Promise.resolve({ data: {}, values: {}, text: '' } as State)),
    createMemory: mock(() => Promise.resolve('test-memory-id' as UUID)),
    actions: [],
    providers: [],
    evaluators: [],
    services: new Map(),
    db: null,
    plugins: [],
    routes: [],
    logger,
    character: {
      name: 'Test SAM Agent',
      id: '00000000-0000-0000-0000-000000000001' as UUID,
      username: 'test-sam-agent',
      bio: ['Retro voice synthesis agent'],
      settings: {},
      system: 'You are a helpful assistant with retro voice synthesis.',
      plugins: ['@elizaos/plugin-sam'],
    },
    ...overrides,
  } as IAgentRuntime;
}

/**
 * Set up logger spies for testing
 */
export function setupLoggerSpies() {
  spyOn(logger, 'info');
  spyOn(logger, 'error');
  spyOn(logger, 'warn');
  spyOn(logger, 'debug');
}

/**
 * Create a mock memory object for testing
 */
export function createMockMemory(overrides: Partial<Memory> = {}): Memory {
  return {
    id: '00000000-0000-0000-0000-000000000002' as UUID,
    entityId: '00000000-0000-0000-0000-000000000003' as UUID,
    agentId: '00000000-0000-0000-0000-000000000001' as UUID,
    roomId: '00000000-0000-0000-0000-000000000004' as UUID,
    content: {
      text: 'test message',
    },
    createdAt: Date.now(),
    ...overrides,
  };
}

/**
 * Create a mock state object for testing
 */
export function createMockState(overrides: Partial<State> = {}): State {
  return {
    data: {},
    values: {},
    text: '',
    ...overrides,
  };
}

/**
 * Create a mock audio buffer for testing
 */
export function createMockAudioBuffer(length: number = 1000): Uint8Array {
  const buffer = new Uint8Array(length);
  for (let i = 0; i < length; i++) {
    buffer[i] = Math.floor(Math.random() * 256);
  }
  return buffer;
}

/**
 * Create a mock WAV buffer with proper header
 */
export function createMockWAVBuffer(audioData: Uint8Array, sampleRate: number = 22050): Uint8Array {
  const wavHeaderSize = 44;
  const wavBuffer = new Uint8Array(wavHeaderSize + audioData.length);

  // WAV header
  const view = new DataView(wavBuffer.buffer, 0, wavHeaderSize);

  // RIFF header
  wavBuffer.set(new TextEncoder().encode('RIFF'), 0);
  view.setUint32(4, 36 + audioData.length, true);
  wavBuffer.set(new TextEncoder().encode('WAVE'), 8);

  // fmt chunk
  wavBuffer.set(new TextEncoder().encode('fmt '), 12);
  view.setUint32(16, 16, true); // fmt chunk size
  view.setUint16(20, 1, true); // PCM format
  view.setUint16(22, 1, true); // Mono
  view.setUint32(24, sampleRate, true);
  view.setUint32(28, sampleRate, true); // Byte rate
  view.setUint16(32, 1, true); // Block align
  view.setUint16(34, 8, true); // 8-bit samples

  // data chunk
  wavBuffer.set(new TextEncoder().encode('data'), 36);
  view.setUint32(40, audioData.length, true);

  // Audio data
  wavBuffer.set(audioData, wavHeaderSize);

  return wavBuffer;
}
